### ISSUE TITLE
250828 Computer Vision Basic_김시진

### ISSUE THREAD

**Issue by kimsijin33** [AIBootcamp14/wiki-rider#37]
## 📅 학습 날짜  
*2025-08-28*

## 📌 키워드 및 요약 정리
키워드: Image Classification, Object Detection, 2-stage Detector, 1-stage Detector, IoU, RoI, AP, YOLO

### - **용어 정의**: 
- **Image Classification**은 컴퓨터 비전 분야에서 대중적인 task이며, CNN과 같은 딥러닝 모델을 사용하여 패턴을 식별하고 이미지를 미리 정의된 클래스로 분류한다. 
- Object Detection은 사물 각가의 Bounding Box(Bbox)위치와 Category를 예측한다. 
- 2-stage Detector는 Region Proposals 및 Feature Extractor를 거치며 Object Detection을 수행
- Region Proposals: 다양한 크기와 모양의 Bounding Box로 물체의 위치를 제안
- Feature Extractor: 제안한 Region (Bounding Box)에 대하여 물체의 특성 추출
- 1-stage Detector는 Region Proposals 없이 Feature Extractor만을 이용한 object detection 수행
- RoI(Region of Interests): 이미지내에서 어떤 특정 부분에 대해 관심을 가지고 작업하는 것을 의미
- IoU(Intersectoin of Union): 우리가 예측한 Bounding Box와 실제 Bounding Box 값의 차이로 성능을 평가하는 성능평가 방법의 하나
- AP(Average Precision): IoU 임계점을 통해 박스의 일치 여부를 판단하여 TP, FP, TN, FN을 구하고 이를 이용해 Precision과 Recall을 계산하여 구함
- YOLO(You Only Look Once): CVPR 2016에 출판된 논문으로, 1-stage Detector 분야의 초기 모델이며 합성곱 신경망을 기반으로 한 일련의 객체 탐지 시스템임. 이미지를 한 번만 보면 객체를 검출할 수 있다 하여 이렇게 이름 붙여짐

### - **핵심 원리**:   
- Image Classification은 CNN 등의 딥러닝 모델로 라벨링 된 이미지의 패턴이나 특징을 학습하여 모델을 만든다. 이 모델은 학습된 패턴을 새로운 이미지에 적용하여 가장 적합한 클래스를 결정한다.
 
### - **시각적/직관적 설명**: 
- 개, 고양이 등의 이미지 분류와 MNIST dataset의 0부터 9까지의 10개로 라벨링 된 숫자 이미지 분류

<img width="1208" height="524" alt="Image" src="https://github.com/user-attachments/assets/9a9745ac-5b11-4a64-b29e-23b72550d64f" />

<img width="1001" height="321" alt="Image" src="https://github.com/user-attachments/assets/d5b25d6d-28fd-44e4-bc9b-ff8fa633f14e" />

 
### - **실습 예시/코드 간략 요약**: 
[CV Basic] (6-2) Image classification inference 해보기.ipynb  코드 요약
1. 개별 Image classification 추론
!wget을 통해 클래스 목록과 강아지 이미지 한 장을 불러옴 
torchvision 라이브러리의 transforms(T) 함수로 이미지를 전처리 함
torch.hub를 통해 학습 된 모델을 불러옴
model.eval() 을 통해 모델을 추론모드로 변경
torch.topk 함수를 통해 logits 에서 가장 큰 값 5개를 추출
torch.softmax 함수를 이용하여 클래스 예측 확률을 표시

2. 여러 장의 Image classification 추론
!wget을 통해 coco  dataset 이미지 12장 불러옴
matplotlib 라이브러리를 통해 이미지 시각화
torchvision 라이브러리의 transforms(T) 함수로 이미지를 전처리함
model.eval() 을 통해 모델을 추론모드로 변경
ground_truth 레이블 인덱스와 레이블 이름 리스트 변수를 만들고 모델의 예측 결과를 확인함
sklearn 내 confusion_matrix, ConfusionMatrixDisplay 함수를 통해 confusion matrix(혼동행렬) 계산하고 시각화함

### 1. 학습 프로세스 

1. Training Data 불러옴 
2. 불러온 Training Data를 모델에서 학습하기 위한 Preprocessing(전처리) 과정을 진행
3. 모델 학습 안정성을 증대시키기 위해 Batch Normalization, Dropout 수행하고 Softmax Classifier로 클래스별 확률 분포 생성 및 모델을 학습시키는데 사용
4. Loss Function은 실제 class와 예측한 class의 차이를 줄이기 위해 사용

### 2. 테스트 프로세스

1. 테스트 데이터 불러옴
2. 불러온 테스트 데이터를 모델에서 예측할 수 있는 데이터로 전 처리 진행
3. 학습된 모델에서 Softmax classfier 등 활성화 함수를 이용하여 클래스별 확율을 예측
4. Prediction 수행 Softmax Classifier의 output 중 가장 큰 값임

### 3. Metric 평가지표

- 모델이 잘 학습되었는지 판별하기 위해 정량적인 평가로 Accuracy와 Precision Metric을 이용

- 수식:

$$Accuracy = TP + TN / TP + TN + FP + FN$$

$$Precision = TP / TP + FP$$ 

- Confusion Matrix(혼돈 행렬)
<img width="430" height="294" alt="Image" src="https://github.com/user-attachments/assets/5436cf7c-c444-463e-9db9-15204b0616ae" />

## ❓ 어려웠던 부분  
- 강의에서 익숙하지 않은 용어들이 조금 나와서 찾아보면서 공부하였고 이미 들어본 용어들도 다시 복습하였습니다. 
예를 들어 2-stage Detector 사용해보기 강의에서 scoop은 논문 스쿱 이나 새로운 기술을 가져다 쓰는 것을 말씀하시는 거 같고, 코드 실습 할 때 plot 을 플로팅 이라고 발음 하셔서 좀 헷갈렸습니다.

- 1-stage Detector 강의 등에서 나오는 수식들을 이해하려고 노력 하였습니다.
익숙하지 않은 기호들이나 변수들을 이해하기 위해 노력하였습니다. 이 부분은 더 공부하겠습니다.

- 혼돈행렬의  TP, TN, FP, FN, Precision. Recall 등은 사실 익숙하지만 아직도 헷갈리는 거 같습니다. 
헷갈리지 않게 더 공부하겠습니다.

## 💬 질문 또는 토론 유도  
- 현재 YOLO11이 YOLO의 최신 버전이고 YOLOv3은 상용화가 가능하다거나 개발하기 좋은 모델이라고 들었는데 맞는 건지 궁금합니다.  

## 🔗 참고 자료 (선택)  
- YOLO - [You Only Look Once](https://ko.wikipedia.org/wiki/%EC%9C%A0_%EC%98%A8%EB%A6%AC_%EB%A3%A9_%EC%9B%90%EC%8A%A4)
- Object Detection - https://wikidocs.net/book/7721
- Batch Normalization(배치 정규화) - https://eehoeskrap.tistory.com/430
- Softmax Classifier - https://aiclaudev.tistory.com/27
- PASCAL VOC - https://docs.ultralytics.com/ko/datasets/detect/voc/
- yolo 버전 - https://docs.ultralytics.com/ko/#yolo-a-brief-history

**Comment by LeeHyunJong** [comment]
분류 문제에 대해서 잘 정리해주셨습니다! Object detection은 용어 정리만 되어있는데, 활용 예시가 다양하므로 추가적으로 정리한다면 도움이 될 것이라고 생각합니다.

### 어려웠던 부분

> 1-stage Detector 강의 등에서 나오는 수식들을 이해하려고 노력 하였습니다. 익숙하지 않은 기호들이나 변수들을 이해하기 위해 노력하였습니다. 이 부분은 더 공부하겠습니다.

YOLO 논문을 한 번 살펴보시면 많은 도움이 될 것입니다.
- $i$: $i$-th grid cell
- $j$: $j$-th bounding box
- $\mathbf{1}_{ij}^{\text{obj}}$: 이 indicator가 1이 되는 경우는 $i$-th grid cell에 객체가 존재하고, $j$-th bounding box가 실제 객체와 IOU가 가장 큰 bounding box일 때입니다. 나머지 경우는 0이 됩니다. 즉, 객체가 존재하는 grid cell에서 예측된(IOU가 가장 높은) bounding box에 대해서 손실함수 1~3번째 항을 계산합니다.
- $(x_i, y_i)$: $i$-th grid cell에서 실제 bounding box의 center / $(\hat{x}, \hat{y})$: 예측값
- $(w_i, h_i)$: $i$-th grid cell에서 실제 bounding box의 width, height / $(hat{w}, \hat{h})$: 예측값
- $C_i$: $i$-th grid cell에서 confidence score; $\mathbf{1}_{ij}^{\text{obj}}=1$일 때, $C_i=$(IOU of $j$-th bounding box \& ground truth). 나머지 경우는 $C_i=0$.
- $\mathbf{1}\_{ij}^{\text{noobj}}$: 1 - $\mathbf{1}\_{ij}^{\text{obj}}$.
- $\mathbf{1}_{i}^{\text{obj}}$: $i$-th grid cell에 객체가 존재할 때 1, 나머지 경우는 0.
- $p_i(c)$: $i$-th grid cell에 객체 c가 존재할 확률 / $\hat{p}_i(c)$: 예측 확률

> 혼돈행렬의 TP, TN, FP, FN, Precision. Recall 등은 사실 익숙하지만 아직도 헷갈리는 거 같습니다. 헷갈리지 않게 더 공부하겠습니다.

종종 헷갈리는 개념인데, TP ~ FN 정의가 예측값을 기준으로 한다는 점을 생각합니다. Precision은 맞다고 예측한 값 중 실제로 맞은 비율, Recall은 실제로 맞는 것 중 내가 제대로 예측한 비율을 의미합니다.

### 질문

> 현재 YOLO11이 YOLO의 최신 버전이고 YOLOv3은 상용화가 가능하다거나 개발하기 좋은 모델이라고 들었는데 맞는 건지 궁금합니다.

자주 쓰인다고 언급드렸던 모델로는 YOLOv5, YOLOv8, YOLOv11이 있습니다.
- YOLOv5는 객체 탐지 전용 모델로, 속도와 안정성이 뛰어나 많이 활용됩니다. 
- YOLOv8은 객체 탐지 외에도 자세 추정 등 다양한 기능을 제공하여 확장성이 높습니다. 
- YOLOv11은 최신 버전으로 SOTA 성능을 보여주지만, 활용 사례가 아직은 많지 않은 것 같습니다.

따라서 실제 프로젝트에서는 사용 환경(메모리, 라이브러리 버전, 경량화 필요 여부 등)을 고려하여 어떤 버전이 적합한지 선택하는 것이 중요하다고 생각합니다.

**Comment by jphan32** [comment]
## 보완 사항
- **Accuracy** 식은 `(TP + TN) / (TP + TN + FP + FN)`처럼 분자와 분모 전체를 괄호로 묶어 주세요. 덧셈 순서를 헷갈리기 쉽습니다. 또, LaTex로 작성하면 분자/분모를 보다 정확하게 작성할 수 있습니다.
- **1-stage Detector**는 **Region Proposal**이 없을 뿐, **Feature Extractor**만으로 끝나지 않고 grid 위에서 클래스와 박스를 동시에 예측하는 **detection head**가 붙습니다. `Backbone → Neck(FPN) → Detection Head` 구성까지 적어 주면 더 정확합니다.

## 보충 설명
- Precision/Recall이 헷갈릴 때는 “모델이 positive라고 판단한 대상”과 “실제 positive” 두 축으로 구분해 도식화해 보세요. 모델 기준/실제 기준으로 나누면 각각의 의미가 분명해집니다.
- 2-stage는 RPN이 후보 영역을 만들고, 1-stage는 이미지 전역 grid에서 바로 분류·박스 회귀를 함께 수행합니다. 차이를 이렇게 정리해 두면 수식이 왜 다른지 자연스럽게 연결됩니다.

## 질문에 대한 답변
- 2024년 기준 공식 배포는 YOLOv8까지이고, 연구 커뮤니티에서는 YOLOv9·YOLOv10 같은 변형이 발표된 상태입니다. YOLOv11이라는 이름은 아직 안정 릴리스가 아니니 출처를 확인해 주세요.
- YOLOv3는 구조가 단순하고 학습 파이프라인이 가벼워 Edge 디바이스에서 여전히 쓰입니다. 다만 최신 데이터셋에서는 anchor 자동 설정, augmentation이 잘 갖춰진 YOLOv5 이상이 더 나은 baseline이 되는 편이라 실험으로 비교해 보세요.

## 학습 팁
- COCO subset으로 YOLOv5n, YOLOv8n을 직접 돌려 mAP과 FPS를 비교해 보면 버전 차이를 체감할 수 있습니다.
- Detector 수식은 실제 anchor 좌표를 손으로 계산해 보면서 파라미터가 어떻게 변하는지 확인해 보세요.
- Confusion matrix와 PR 커브를 같이 그려 두면 어떤 클래스에서 오류가 큰지 바로 확인할 수 있습니다.
